cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -205,23 +210,74 @@ pub fn legalize_signature(
         }
     }
 
-    legalize_args(&mut sig.params, &mut args);
-
-    let (regs, fpr_limit) = if sig.call_conv.extends_windows_fastcall() {
+    let (ret_regs, ret_fpr_limit) = if sig.call_conv.extends_windows_fastcall() {
         // windows-x64 calling convention only uses XMM0 or RAX for return values
         (&RET_GPRS_WIN_FASTCALL_X64[..], 1)
     } else {
         (&RET_GPRS[..], 2)
     };
-
     let mut rets = Args::new(
         bits,
-        regs,
-        fpr_limit,
+        ret_regs,
+        ret_fpr_limit,
         sig.call_conv,
         shared_flags,
         isa_flags,
     );
+
+    if sig.returns.len() > 1 {
+        // We're using the first register for the return pointer parameter.
+        let mut ret_ptr_param = AbiParam {
+            value_type: args.pointer_type,
+            purpose: ArgumentPurpose::StructReturn,
+            extension: ArgumentExtension::None,
+            location: ArgumentLoc::Unassigned,
+        };
+        match args.assign(&ret_ptr_param) {
+            ArgAction::Assign(loc) => {
+                ret_ptr_param.location = loc;
+                sig.params.insert(0, ret_ptr_param);
+            }
+            ArgAction::Convert(_) => unreachable!(),
```

Comment:
Is it correct to phrase this as ``unreachable!("`pointer_type` shouldn't require conversion")``?